dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -370,124 +432,71 @@ private void handlePotentialFormatEntry(Set<FileFormat> fileFormats, Object v) {
+                    stepDockerPath = getRequirementOrHint(expressionTool.getRequirements(), expressionTool.getHints(), stepDockerPath);
+                    stepToType.put(workflowStepId, EXPRESSION_TOOL_TYPE);
+                    entryId = convertToString(expressionTool.getId());
                 } else {
-                    nodeDockerInfo.put(workflowStepId, new DockerInfo(mainDescriptorPath, stepDockerRequirement, dockerUrl, dockerSpecifier));
+                    LOG.error(CWLHandler.CWL_PARSE_SECONDARY_ERROR + run);
+                    throw new CustomWebApplicationException(CWLHandler.CWL_PARSE_SECONDARY_ERROR + run, HttpStatus.SC_UNPROCESSABLE_ENTITY);
                 }
-
+                currentPath = preprocessor.getPath(entryId);
+            } else {
+                stepToType.put(workflowStepId, "n/a");
+                currentPath = run.toString();
             }
 
-            if (type == LanguageHandlerInterface.Type.DAG) {
-                // Determine steps that point to end
-                List<String> endDependencies = new ArrayList<>();
-
-                for (WorkflowOutputParameter workflowOutputParameter : workflow.getOutputs()) {
-                    Object sources = workflowOutputParameter.getOutputSource();
-                    processDependencies(nodePrefix, endDependencies, sources);
-                }
-
-                toolInfoMap.put("UniqueEndKey", new ToolInfo(null, endDependencies));
-                nodePairs.add(new MutablePair<>("UniqueEndKey", ""));
+            if (currentPath == null) {
+                currentPath = "";
+            }
 
-                // connect start node with them
-                for (Pair<String, String> node : nodePairs) {
-                    if (toolInfoMap.get(node.getLeft()) == null) {
-                        toolInfoMap.put(node.getLeft(), new ToolInfo(null, Lists.newArrayList("UniqueBeginKey")));
-                    }
-                }
-                nodePairs.add(new MutablePair<>("UniqueBeginKey", ""));
+            DockerSpecifier dockerSpecifier = null;
+            String dockerUrl = null;
+            String stepType = stepToType.get(workflowStepId);
+            if ((stepType.equals(WORKFLOW_TYPE) || stepType.equals(TOOL_TYPE)) && !Strings.isNullOrEmpty(stepDockerPath)) {
+                // CWL doesn't support parameterized docker pulls. Must be a string.
+                dockerSpecifier = LanguageHandlerInterface.determineImageSpecifier(stepDockerPath, DockerImageReference.LITERAL);
+                dockerUrl = getURLFromEntry(stepDockerPath, dao, dockerSpecifier);
+            }
 
-                return Optional.of(setupJSONDAG(nodePairs, toolInfoMap, stepToType, nodeDockerInfo));
-            } else {
-                return Optional.of(getJSONTableToolContent(nodeDockerInfo));
+            if (depth == 0 && type == LanguageHandlerInterface.Type.DAG) {
```

Comment:
This will be followed up on in another ticket, right?

Reply:
Yes, there is an existing DAG ticket:  https://ucsc-cgl.atlassian.net/browse/DOCK-1282 https://github.com/dockstore/dockstore/issues/3352